File: 21/day4.py
chosen = []
numeros = {}
matrizes = []

f = open("exinput.txt", "r")

matrizAtual = []
linhaAtual = []

inputLines = linhas = matriz = 0

for line in f:
    inputLines+=1
    if inputLines == 1:
        chosen = list(map(int, line[:-1].split(",")))
        inputLines+=1
        continue
    
    linha = line[:-1].split()
    if linha:
        for idx, num in enumerate(linha):
            linhaAtual.append({int(num): 0})
            if int(num) in numeros:
                numeros[int(num)].append((matriz,linhas,idx))
            else:
                numeros[int(num)] = [(matriz,linhas, idx)] 

        linhas+=1

    else:
        if matrizAtual:
            matrizes.append(matrizAtual)
            matrizAtual = []
            linhas = 0
            matriz+=1
        continue

    matrizAtual.append(linhaAtual)
    linhaAtual = []

bingo = False
for num in chosen:
    for i in range(len(numeros[num])):
        values = numeros[num][i]
        z = values[0] # which matrix
        x = values[1] # which line
        y = values[2] # which column
        matrizes[z][x][y][num] = 1

        pontosLinha = sum(list(value.values())[0] for value in matrizes[z][x])

#        if(pontosLinha == 5):
#            bingo = True
#            print("BINGOOO: ", matrizes[z][x])

        pontosColuna = 0 
        for lin in range(5):
            pontosColuna += list(matrizes[z][lin][y].values())[0]
        
        if pontosColuna == 5 or pontosLinha == 5:
#            for value in matrizes[z]:
#                print(value)
#            for i in range(5):
#                for j in range(5):
#                    print(list(matrizes[z][i][j].values())[0])
#                    if matrizes[z][i][j]:

            soma = 0
            for line in matrizes[z]:
                for dic in line:
                    valor = list(dic.values())[0]
                    if valor == 0:
                        soma += list(dic.keys())[0]
            print("soma = ", soma, " num = ", num, " total = ", soma*num )
            print("part 1 = ", sum(map(sum, ((value.values())[0] for value in matrizes[z]))))
            print(matrizes)
            exit()

File: 21/day4.c
#include <stdio.h>
#include <stdlib.h>

int chosen[150];

int matrix[100][5][5];
int sorted[100][5][5] = {0};

int main(){

    FILE* file = fopen ("exinput.txt", "r");

    if (file == NULL){
        printf("Error! Couldnt open the file!");
        exit(1);
    }

    int flag = 0, lin = 0, col = 0, mat = 0, num = 0;
    while(1){
        int tmp;
        if( fscanf(file, "%d,", &tmp) != 1 ) break;

        // change to last digit from input chosen numbers
        if( !flag ){
            chosen[lin++] = tmp;
            if(tmp == 1) {flag = 1; lin = 0;}
            num++;
            continue;
        }
        
        // 5 x 5 matrix
        if( col == 5 ){
            col = 0;
            lin++;
            if ( lin == 5 ){
                lin = 0;
                mat++;
            }
        }

        matrix[mat][lin][col++] = tmp;
    }

    int bingo = 0;
    for(int z = 0; z < num; z++){
        int sort = chosen[z];

        printf("\n---new number %d---\n", sort);

        for(int i = 0; i < 3; i++ ){

            int markedColumn[5] = {0};
            for( int j = 0; j < 5; j++ ){

                int flag = 0;
                int markedLine = 0;
                for( int k = 0; k < 5; k++ ){
                    if( sort == matrix[i][j][k] ) sorted[i][j][k] = 1;
                    if( sorted[i][j][k] == 1) { markedLine++; if( flag == 0 ) {markedColumn[j]++; flag = 1;}}
                    printf("%d ", sorted[i][j][k]);
                }
                printf("\n");

                printf("marked column = %d %d %d %d %d\n", markedColumn[0], markedColumn[1], markedColumn[2], markedColumn[3], markedColumn[4]);

                if( markedLine == 5 ){
                    printf("BINGO!!\n");
                    exit(1);
                }
            }
            printf("\n");
        }
    }
    return 0;
}
